## Eventos INM a Nivel municipio

Notebook para trabajar datos de encuentros con migrantes a nivel municipal.
El objetivo de este notebook es extraer dos documentos;

1)Archivo Geoespacial con los boundaries municipales para implementar en arcgis

2)Archivo csv + anexos geoespaciales para generar mapas con proyecciones estadisticas

### Eventos de extranjeros presentados ante la autoridad migratoria, según municipio.

Procesamos las tablas estadisticas con datos sobre eventos de migrantes irregulares registrados por el INM por entidad federativa.

Los datos seleccionados se refieren unicamente a los eventos en los que migrantes fueron  ...

    2001 - 2012 la información incluye eventos de extranjeros en situación migratoria irregular a los que el Instituto Nacional de Migración (INM) les inició un procedimiento administrativo de aseguramiento y a los centroamericanos acogidos al “Memorándum de entendimiento entre los gobiernos de los Estados Unidos Mexicanos, de la República de El Salvador, de la República de Guatemala, de la República de Honduras y de la República de Nicaragua, para la repatriación digna, ordenada, ágil y segura de nacionales centroamericanos migrantes vía terrestre” suscrito el 5 de mayo de 2006 y su anexo del 26 de abril de 2007.

    2013 - 2020 la información incluye los eventos de migrantes en situación migratoria irregular a los que se les inició un procedimiento administrativo de presentación por no acreditar su situación migratoria, según lo previsto en los arts. 99, 112 y 113 de la Ley de Migración y del art. 222 de su Reglamento.

    2021 - 2023 la información hace referencia a los eventos de extranjeros en situación migratoria irregular a los que se les inicio un Procedimiento Administrativo Migratorio (PAM) ante el Instituto Nacional de Migración (INM) por no acreditar su situación migratoria, según lo previsto en los arts. 99, 100, 101 y 113 de la Ley de Migración y del art. 222 de su Reglamento 

Y a los que fueron canalizados por esta misma autoridad a los albergues de la red DIF, con PAM iniciado; según lo previsto en los art. 112 y 113 de la Ley de Migración y del art. 222 de su Reglamento, así como de los arts. 89, 94 de la Ley General de los Derechos de Niñas, Niños y Adolescentes.(Excluidos en estas BD para mantener comparabilidad estadistica)

El nombre de datos en el tiempo:

2002 - 2006 "EVENTOS DE ASEGURAMIENTO EN MÉXICO SEGÚN DELEGACIÓN REGIONAL"

2007 - 2012  "Eventos de extranjeros alojados en estaciones migratorias, según entidad federativa"

2013 - 2023 "Eventos de extranjeros presentados ante la autoridad migratoria, según entidad federativa"



# Introducción
Este notebook analiza la dinámica de flujos migratorios irregulares en México, utilizando herramientas de análisis espacial como GeoPandas.

Fuente: http://www.politicamigratoria.gob.mx/es/PoliticaMigratoria/Boletines_Estadisticos

## Configuración Inicial e Importaciones

In [1]:
#Librerias Necesarias
import pandas as pd
import glob
import numpy as np
import seaborn as sns
from unidecode import unidecode
import re
import os
import sys
import geopandas as gpd
from geodatasets import get_path


from geopy.geocoders import Photon
from geopy.exc import GeocoderTimedOut
from shapely.geometry import Point


import matplotlib.pyplot as plt
import squarify
import seaborn as sns
import matplotlib as mpl

import plotly.express as px
import plotly.graph_objects as go

## Preparación de los Datos

### Carga de Datos INM Municipal (Archivo no geoespacial)

La limpieza y prepraración de set de datos Incluye:

    -1 revisar el archivo desde excel para remover todas las columnas que corresponden a titulos y notas.
    -2 cargar la bd en python
    -3 deberemos normalizar las columnas para que correspondan a un mes.
    en el caso de los datos posteriores al año 2023 los cuales tienen columnas a dos niveles: mes (presentados, canalizados y subtotal), deberemos combinar las columnas con cada mes: "enero_presentados", "enero_canalizados", "enero_subtotal"

In [2]:
# Ruta al archivo .xls
archivo_xls = '/Users/pablouriarte/Documents/1. Expediente Tec de Monterrey/1.Tesis/Mapa_Migracion_Irregular_Mexico/2. datos_estadisticos/1. Detenciones INM/Total_De_Eventos/Eventos Municipal/3.1_mun_2023.xlsx'

# Cargar el archivo xls como un DataFrame de pandas
datos = pd.read_excel(archivo_xls)

Eliminar columnas y filas innecesarias: Podemos quitar las columnas y filas que solo contienen valores NaN o que no aportan información útil para el análisis.

In [3]:
# Eliminar las columnas completamente vacías
datos = datos.dropna(axis=1, how='all')

# Eliminar las filas completamente vacías
datos = datos.dropna(axis=0, how='all')

Renombrar las columnas: Las columnas deben tener nombres descriptivos para facilitar el entendimiento y el acceso a los datos.

In [4]:
# Establecer la primera fila como nombres de columna
datos.columns = datos.iloc[0]

# Descartar la primera fila del DataFrame
datos = datos.drop(datos.index[0])

# Resetea el índice si es necesario
datos = datos.reset_index(drop=True)

In [5]:
#lista de columnas:
lista_columnas = datos.columns.tolist()
print(lista_columnas)

['Entidad Federativa / Municipio', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Subtotal', 'Presentados', 'Canalizados', 'Total']


Sabemos del df original que cada set de 3 columans corresponde a 1 mes despues de la columna [0], por lo que es mas facil renombrarlas:

In [6]:
# Nuevo mapeo de nombres para las columnas
nuevos_nombres = [
    'Entidad Federativa / Municipio',
    'enero_presentados', 'enero_canalizados', 'enero_total',
    'febrero_presentados', 'febrero_canalizados', 'febrero_total',
    'marzo_presentados', 'marzo_canalizados', 'marzo_total',
    'abril_presentados', 'abril_canalizados', 'abril_total',
    'mayo_presentados', 'mayo_canalizados', 'mayo_total',
    'junio_presentados', 'junio_canalizados', 'junio_total',
    'julio_presentados', 'julio_canalizados', 'julio_total',
    'agosto_presentados', 'agosto_canalizados', 'agosto_total',
    'septiembre_presentados', 'septiembre_canalizados', 'septiembre_total',
    'octubre_presentados', 'octubre_canalizados', 'octubre_total',
    'noviembre_presentados', 'noviembre_canalizados', 'noviembre_total',
    'diciembre_presentados', 'diciembre_canalizados', 'diciembre_total',
    'total_presentados', 'total_canalizados', 'total_total'
]

# Asignación de nuevos nombres a las columnas del DataFrame
datos.columns = nuevos_nombres




In [13]:
datos.head()

Entidad Federativa / Municipio       tipo enero_presentados  \
0                  Total general     estado             24801   
1                 Aguascalientes     estado               153   
2           Aguascalientes, Ags.  municipio               153   
3                    Cosío, Ags.  municipio                 0   
4              Jesús María, Ags.  municipio                 0   

  enero_canalizados enero_total febrero_presentados febrero_canalizados  \
0             12559       37360               23631               14410   
1                13         166                 155                  19   
2                13         166                 155                  19   
3                 0           0                   0                   0   
4                 0           0                   0                   0   

  febrero_total marzo_presentados marzo_canalizados  ...  \
0         38041             31360             13268  ...   
1           174                17                13  ...   
2           174                15                12  ...   
3             0                 2                 1  ...   
4             0                 0                 0  ...   

  noviembre_presentados noviembre_canalizados noviembre_total  \
0                 72789                 25180           97969   
1                     3                    25              28   
2                     0                     0               0   
3                     1                    20              21   
4                     2                     5               7   

  diciembre_presentados diciembre_canalizados diciembre_total  \
0                 79946                 15507           95453   
1                     0                    16              16   
2                     0                     0               0   
3                     0                    16              16   
4                     0                     0               0   

  total_presentados total_canalizados total_total  \
0            566361            215815      782176   
1               386               173         559   
2               348                75         423   
3                27                71          98   
4                 8                12          20   

                           ciudad  
0                   Total general  
1                  Aguascalientes  
2  Aguascalientes, Aguascalientes  
3           Cosío, Aguascalientes  
4     Jesús María, Aguascalientes  

[5 rows x 42 columns]

creamos nueva columna ciudad, la cual contiene valores: "municipio, estado"

In [10]:
# Crea la columna 'tipo' basada en la presencia de una coma en 'Entidad Federativa / Municipio'
datos['tipo'] = datos['Entidad Federativa / Municipio'].apply(lambda x: 'municipio' if ',' in x else 'estado')

# Mueve la columna 'tipo' a la posición deseada, que es la segunda posición (índice 1)
# Primero, guardamos la columna 'tipo' y luego la eliminamos de su posición original
columna_tipo = datos['tipo']
datos.drop(columns=['tipo'], inplace=True)

# Luego, insertamos la columna en la posición deseada
datos.insert(1, 'tipo', columna_tipo)

In [12]:
# Primero, inicializa la columna 'ciudad' con valores vacíos.
datos['ciudad'] = ''

# Almacenar el estado actual mientras iteramos.
estado_actual = None

# Iterar sobre las filas del dataframe.
for index, row in datos.iterrows():
    if row['tipo'] == 'estado':
        # Cuando encuentra un 'estado', actualiza el estado actual y asigna el nombre del estado a la columna 'ciudad'.
        estado_actual = row['Entidad Federativa / Municipio']
        datos.at[index, 'ciudad'] = estado_actual
    elif row['tipo'] == 'municipio' and estado_actual:
        # Para los 'municipio', combina el nombre del municipio con el estado actual.
        ciudad = row['Entidad Federativa / Municipio'].split(',')[0]  # Asumiendo que el nombre de la ciudad está antes de la coma.
        datos.at[index, 'ciudad'] = f"{ciudad}, {estado_actual}"


Hacemos un subdf con los municipios, y dropeamos las columnas que ya no necesitamos

In [14]:
# Filtrado para obtener solo las filas donde 'tipo' es igual a 'municipio'
datos_municipio = datos.loc[datos['tipo'] == 'municipio']

In [16]:
# Mover la columna 'ciudad' al principio del dataframe
col_order = ['ciudad'] + [col for col in datos_municipio.columns if col not in ['ciudad', 'Entidad Federativa / Municipio', 'tipo']]
datos_municipio = datos_municipio[col_order]


In [18]:
datos_municipio.shape

(389, 40)

In [28]:
datos_municipio.head()

ciudad   Latitude   Longitude  \
2             Aguascalientes, Aguascalientes  22.000000 -102.500000   
3                      Cosío, Aguascalientes  22.370515 -102.312009   
4                Jesús María, Aguascalientes  21.934556 -102.468452   
5  San Francisco de los Romo, Aguascalientes  22.074628 -102.270713   
7                  Ensenada, Baja California  31.865889 -116.602983   

                              Coordinates enero_presentados enero_canalizados  \
2                 22.0000001,-102.5000001               153                13   
3          22.3705155,-102.31200851052341                 0                 0   
4  21.934555500000002,-102.46845171651151                 0                 0   
5                 22.0746282,-102.2707133                 0                 0   
7                  31.8658887,-116.602983                 4                 0   

  enero_total febrero_presentados febrero_canalizados febrero_total  ...  \
2         166                 155                  19           174  ...   
3           0                   0                   0             0  ...   
4           0                   0                   0             0  ...   
5           0                   0                   0             0  ...   
7           4                   1                   0             1  ...   

  octubre_total noviembre_presentados noviembre_canalizados noviembre_total  \
2            30                     0                     0               0   
3            27                     1                    20              21   
4             0                     2                     5               7   
5             3                     0                     0               0   
7             0                     0                     0               0   

  diciembre_presentados diciembre_canalizados diciembre_total  \
2                     0                     0               0   
3                     0                    16              16   
4                     0                     0               0   
5                     0                     0               0   
7                     1                     0               1   

  total_presentados total_canalizados total_total  
2               348                75         423  
3                27                71          98  
4                 8                12          20  
5                 3                15          18  
7                23                 6          29  

[5 rows x 43 columns]

Agregamos una columna geoespacial para hacer la union con la df de los poligonos municipales.

In [19]:
# Inicializar el geolocalizador de Photon
geolocator = Photon(user_agent="geoapiExercises")

# Función para obtener latitud, longitud y coordenadas combinadas
def get_lat_long(address):
    try:
        location = geolocator.geocode(address)
        if location:
            return (location.latitude, location.longitude, f"{location.latitude},{location.longitude}")
        else:
            return (None, None, None)
    except:
        return (None, None, None)

# Aplicar la función a la columna 'Entidad Federativa / Municipio'
datos_municipio['Lat_Long_Coords'] = datos_municipio['ciudad'].apply(get_lat_long)

# Separar las coordenadas en tres columnas nuevas: 'Latitude', 'Longitude' y 'Coordinates'
datos_municipio[['Latitude', 'Longitude', 'Coordinates']] = pd.DataFrame(datos_municipio['Lat_Long_Coords'].tolist(), index=datos_municipio.index)

# Eliminar la columna 'Lat_Long_Coords' ya que se ha dividido en otras columnas
datos_municipio.drop(columns=['Lat_Long_Coords'], inplace=True)

In [25]:
last_three_cols = datos_municipio.columns[-3:].tolist()

# Define el nuevo orden de las columnas: 'ciudad', seguido por las últimas tres columnas, y luego todas las demás, excepto las últimas tres.
new_col_order = ['ciudad'] + last_three_cols + [col for col in datos_municipio.columns if col not in ['ciudad'] + last_three_cols]

# Reordena las columnas usando el nuevo orden
datos_municipio = datos_municipio[new_col_order]

### Datos Estadísticos

### Geolocalizaciones: Instituciones

##### Cargado Limites Municipales de Mexico

In [22]:
# Mex Municipios
mx_mun = gpd.read_file('/Users/pablouriarte/Documents/1. Expediente Tec de Monterrey/1.Tesis/Mapa_Migracion_Irregular_Mexico/3. mapas/2. arcgis/maps_boundaries/Municipalities/mapa_mexico_mun/').set_index('CLAVE').to_crs(epsg=4485)

In [23]:
mx_mun.shape

(2480, 6)

Trabajaremos aada municipio uniendolo con la columna clave

In [ ]:
mx_mun.head()

In [ ]:
print(mx_mun.iloc[0, mx_mun.columns.get_loc('geometry')])


unir las bds:

### Union de ambas bd.

-utilizando los datos geoespaciales para hacer el merge

## Archivos para descarga

#### Pruebas: '/Users/pablouriarte/Documents/1. Expediente Tec de Monterrey/1.Tesis/Mapa_Migracion_Irregular_Mexico/2. datos_estadisticos/1. Detenciones INM/Total_De_Eventos/Eventos Municipal/3.1_mun_2023_pruebas.xlsx'

### Tabla de datos unida con los centroides (solo puntos, funciona muy bien para hacer heatmaps ) 389 municipios con datos y los poligonos de todos los municipios de municipios.

In [ ]:
# Asegúrate de que 'datos_municipales' es el nombre de tu DataFrame
#3.1_mun_2023_centr.csv
datos_municipio.to_csv('/Users/pablouriarte/Documents/1. Expediente Tec de Monterrey/1.Tesis/Mapa_Migracion_Irregular_Mexico/2. datos_estadisticos/1. Detenciones INM/Total_De_Eventos/Eventos Municipal/3.1_mun_2023_centr.csv', index=False)


### Tabla unida con 389 municipios con datos y los poligonos de todos los municipios de municipios

### Tabla unida centralmente datos y geodatos de solo los 389 municipios con datos

### Distribución Institucional

### Visualización de Flujos Migratorios

### Mapas Interactivos

## Conclusiones

## Referencias y Recursos Adicionales